File: elixir_intro/lib/inc_server.ex
defmodule IncServer do
  use GenServer # wir implementieren ein Interface
  # GenServer gehört zu OTP: Open Telephony Platform

  def start(n) do
    # 1. startet einen neuen Prozeß
    # 2. ruft in diesem Prozeß IncServer.init/1 auf, übergibt dabei n,
    #    bekommt es den initialen Zustand
    # 3. startet dann eine endrekursive Schleife, die Nachrichten akzeptiert,
    #    beantwortet und den Zustand verwaltet
    GenServer.start(__MODULE__, n)
  end

  def init(n) do
    {:ok, n} # n: initialer Zustand
  end

  # GenServer unterscheidet zwischen "fire-or-forget"-Nachrichten: "cast"
  # NICHT WIE JAVA-CAST!
  # wie Inc
  # und Nachrichten, die eine Antwort erwarten: "call"
  # wie Get

  # Messages, die inc_loop akzeptiert:

  # um Wert inkrementieren
  defmodule Inc do
    use QuickStruct, [i: number()]
  end
  # Wert abholen
  defmodule Get do
    use QuickStruct, []
  end

  def handle_cast(%Inc{i: i}, n) do
    IO.puts("current state: #{n}")
    {:noreply, n + i}
  end

  def handle_call(%Get{}, _from, n) do
    # Doku:  {:reply, reply, new_state}
    {:reply,
     n, # Antwort, GenServer schickt die automatisch zurück
     n} # neue Zustand
  end
end

File: erlang/inc_server.erl
-module(inc_server).

-behavior(gen_server).

-export([init/1, start/1,
         inc/2, get/1,
         handle_cast/2, handle_call/3]).

-record(get, {sender_pid :: pid() }). % sender_pid bei gen_server überflüssig
-record(inc, {increment :: number()}).

%% N: Anfangsstand des Zählers
init(N) ->
    {ok, N}. % Anfangszustand vom gen_server

inc(Pid, Increment) ->
    gen_server:cast(Pid, #inc{increment = Increment}).

get(Pid) ->
    gen_server:call(Pid, #get{sender_pid = self()}).

start(N) ->
    gen_server:start({global, global_inc}, ?MODULE, N, []).
                                                  % ^ wird an init übergeben

handle_cast(#inc{increment = Increment}, N) ->
    {noreply, N + Increment}.

handle_call(#get{}, _From, N) ->
    {reply, N, N}.

